### PR TITLE
Allow control plane migration across seeds w/ different provider types

### PR DESCRIPTION
**Pull request by plkokanov**
**How to categorize this PR?**

/area control-plane-migration
/kind enhancement

**What this PR does / why we need it**:
This PR allows migration of Shoot control planes across Seeds with different provider types.
It also adds warnings and workarounds regarding problems that could occur due to lack of network connectivity:
1. When pods in the Destination Seed do not have network connectivity to the storage provider of the Source Seeds, meaning that it will be impossible to automatically copy etcd backups during the `Restore` operation
2. When the Shoot's nodes do not have network connectivity to the control plane after it is moved to the Destination Seed

Additionally, required provider extensions for source `BackupEntry`s will now be correctly installed in the Destination Seed during control plane migration.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature operator
Performing control plane migration across `Seed`s with different provider types is now possible.  Before triggering the migration, make sure that pods in the `Shoot`'s control plane, once it is moved to the `Destination Seed`, will have network connectivity to the storage provider of the `Source Seed` (so that ETCD backups can be copied automatically). Additionally, make sure that the `Shoot`'s nodes will have network connectivity to the `Shoot`'s control plane after it is moved to the `Destination Seed`.
```

### CODE REVIEW

**Review by dimitar-kostadinov — COMMENTED** [review]
/lgtm

File: docs/operations/control_plane_migration.md [hunk 1]
```diff
@@ -2,15 +2,15 @@
...
 
 ## Shoot Control Plane Migration
 
-Triggering the migration is done by changing the `Shoot`'s `.spec.seedName` to a `Seed` that differs from the `.status.seedName`, we call this `Seed` a `"Destination Seed"`. This action can only be performed by an operator with the necessary RBAC. If the Destination `Seed` does not have a backup and restore configuration, the change to `spec.seedName` is rejected. Additionally, this Seed must not be set for deletion and must be healthy.
+Triggering the migration is done by changing the `Shoot`'s `.spec.seedName` to a `Seed` that differs from the `.status.seedName`, we call this `Seed` a `"Destination Seed"`. This action can only be performed by an operator with the necessary RBAC. If the `Destination Seed` does not have a backup and restore configuration, the change to `spec.seedName` is rejected. Additionally, this Seed must not be set for deletion and must be healthy.
```
Comment (timuthy) [inline docs/operations/control_plane_migration.md]:
```suggestion
Triggering the migration is done by changing the `Shoot`'s `.spec.seedName` to a `Seed` that differs from the `.status.seedName`, we call this `Seed` a `"Destination Seed"`. This action can only be performed by an operator (see [Triggering the Migration](#triggering-the-migration)). If the `Destination Seed` does not have a backup and restore configuration, the change to `spec.seedName` is rejected. Additionally, this Seed must not be set for deletion and must be healthy.
```

Nit.

File: docs/operations/control_plane_migration.md [hunk 2]
```diff
@@ -24,11 +24,38 @@ The etcd backups will be copied over to the `BackupBucket` of the `Destination S
...
+>
+> The nodes of your `Shoot` cluster must have network connectivity to the `Shoot`'s `kube-apiserver` and the `vpn-seed-server` once they are migrated to the `Destination Seed`. Otherwise, the `Restore` operation will get stuck at the `Waiting until the Kubernetes API server can connect to the Shoot workers` step. However, if you do end up in this case and cannot allow network traffic from the nodes to the `Shoot`'s control plane, you can annotate the `Shoot` with the `shoot.gardener.cloud/skip-readiness` annotation so that the `Restore` operation finishes, and then use the [`shoots/binding`](../concepts/scheduler.md#shootsbinding-subresource) subresource to migrate the control plane back to the `Source Seed`.
+
+
+## Copying ETCD Backups Manually During the `Restore` Operation
+
+Following is a workaround that can be used to copy etcd backups manually in situations where a `Shoot`'s control plane has been moved to a `Destination Seed` and the pods running in it lack network connectivity to the `Source Seed`s storage provider:
```
Comment (timuthy) [inline docs/operations/control_plane_migration.md]:
```suggestion
Following is a workaround that can be used to copy etcd backups manually in situations where a `Shoot`'s control plane has been moved to a `Destination Seed` and the pods running in it lack network connectivity to the `Source Seed`'s storage provider:
```

File: docs/operations/control_plane_migration.md [hunk 3]
```diff
@@ -24,11 +24,38 @@ The etcd backups will be copied over to the `BackupBucket` of the `Destination S
...
 ```bash
 NAMESPACE=my-namespace
 SHOOT_NAME=my-shoot
 DEST_SEED_NAME=destination-seed
+
 kubectl get --raw /apis/core.gardener.cloud/v1beta1/namespaces/${NAMESPACE}/shoots/${SHOOT_NAME} | jq -c '.spec.seedName = "'${DEST_SEED_NAME}'"' | kubectl replace --raw /apis/core.gardener.cloud/v1beta1/namespaces/${NAMESPACE}/shoots/${SHOOT_NAME}/binding -f - | jq -r '.spec.seedName'
 ```
+
+
+> [!NOTE]
```
Comment (timuthy) [inline docs/operations/control_plane_migration.md]:
```suggestion
> [!IMPORTANT]
```

Nit.

File: docs/operations/control_plane_migration.md [hunk 4]
```diff
@@ -24,11 +24,38 @@ The etcd backups will be copied over to the `BackupBucket` of the `Destination S
...
+1. After you have successfully copied the etcd backups, wait for the `EtcdCopyBackupsTask` custom resource to be created in the `Shoot`'s control plane on the `Destination Seed`, if it does not already exist. Afterwards, mark it as successful by patching it using the following command:
+    ```
+    SHOOT_NAME=my-shoot
+    PROJECT_NAME=my-project
+
+    kubectl patch -n shoot--${PROJECT_NAME}--${SHOOT_NAME} etcdcopybackupstask ${SHOOT_NAME} --subresource status --type merge -p "{\"status\":{\"conditions\":[{\"type\":\"Succeeded\",\"status\":\"True\",\"reason\":\"manual copy successful\",\"message\":\"manual copy successful\",\"lastTransitionTime\":\"$(date -Iseconds)\",\"lastUpdateTime\":\"$(date -Iseconds)\"}]}}"
+    ```
+1. After the `main-etcd` becomes `Ready`, remove the finalizer on the source `extensions.gardener.cloud/v1alpha1.BackupEntry` (the resource has `source-` as a prefix in its name) in the `Destination Seed` so that it can be deleted successfully. This is necessary as the `Destination Seed` will not have network connectivity to perform the actual deletion of the source backup directory.
```
Comment (timuthy) [inline docs/operations/control_plane_migration.md]: > This is necessary as the `Destination Seed` will not have network connectivity to perform the actual deletion of the source backup directory.

In the next step this cleanup ought to happen manually. Doesn't this lead to the proper deletion of the source `BackupEntry`, without removing the finalizer?

**Review by timuthy — COMMENTED** [review]
Thanks 🙂

File: pkg/controllermanager/controller/controllerregistration/seed/reconciler.go [hunk 5]
```diff
@@ -83,9 +84,13 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	}
 
 	backupBucketList := &gardencorev1beta1.BackupBucketList{}
-	if err := r.Client.List(ctx, backupBucketList, client.MatchingFields{core.BackupBucketSeedName: seed.Name}); err != nil {
+	if err := r.Client.List(ctx, backupBucketList); err != nil {
 		return reconcile.Result{}, err
 	}
+	backupBucketsMap := make(map[string]*gardencorev1beta1.BackupBucket, len(backupBucketList.Items))
```
Comment (rfranzke) [inline pkg/controllermanager/controller/controllerregistration/seed/reconciler.go]:
```suggestion
	backupBucketNameToObject := make(map[string]*gardencorev1beta1.BackupBucket, len(backupBucketList.Items))
```

File: pkg/controllermanager/controller/controllerregistration/seed/reconciler.go [hunk 6]
```diff
@@ -155,30 +160,26 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
...
-	backupBucketList *gardencorev1beta1.BackupBucketList,
-) (
-	sets.Set[string],
-	map[string]gardencorev1beta1.BackupBucket,
-) {
-	var (
-		wantedKindTypeCombinations = sets.New[string]()
-		buckets                    = make(map[string]gardencorev1beta1.BackupBucket)
-	)
+	backupBucketsMap map[string]*gardencorev1beta1.BackupBucket,
```
Comment (rfranzke) [inline pkg/controllermanager/controller/controllerregistration/seed/reconciler.go]: same here :)

File: pkg/controllermanager/controller/controllerregistration/seed/reconciler.go [hunk 7]
```diff
@@ -155,30 +160,26 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
...
+	return wantedKindTypeCombinations
 }
 
 // computeKindTypesForBackupEntries computes the list of wanted kind/type combinations for extension resources based on the
 // the list of existing BackupEntry resources.
 func computeKindTypesForBackupEntries(
 	log logr.Logger,
 	backupEntryList *gardencorev1beta1.BackupEntryList,
-	buckets map[string]gardencorev1beta1.BackupBucket,
+	buckets map[string]*gardencorev1beta1.BackupBucket,
```
Comment (rfranzke) [inline pkg/controllermanager/controller/controllerregistration/seed/reconciler.go]: same here :)

File: pkg/gardenlet/controller/backupentry/reconciler.go [hunk 8]
```diff
@@ -356,6 +352,8 @@ func (r *Reconciler) deleteBackupEntry(
 			}
 			log.Info("Extension BackupEntry not yet deleted", "extensionBackupEntry", client.ObjectKeyFromObject(extensionBackupEntry))
 			return reconcile.Result{RequeueAfter: RequeueDurationWhenResourceDeletionStillPresent}, nil
+		} else if !apierrors.IsNotFound(err) {
+			return reconcile.Result{}, err
```
Comment (rfranzke) [inline pkg/gardenlet/controller/backupentry/reconciler.go]: What is the purpose of this change? 👀

Comment (plkokanov) [reply pkg/gardenlet/controller/backupentry/reconciler.go]: The linter was issuing a warning about the old check because in the `else` condition `err` is actually always `nil`
```
if err := r.SeedClient.Get(seedCtx, client.ObjectKeyFromObject(extensionBackupEntry), extensionBackupEntry); err != nil {
			if !apierrors.IsNotFound(err) {
				return reconcile.Result{}, err
			}
		} else if err == nil {
```

However, if I just leave it with `} else {` then the linter issues an error this time that the code could be optimised further - and that's the change I've done here.

Comment (rfranzke) [reply pkg/gardenlet/controller/backupentry/reconciler.go]: Which linter? 👀 I don't see any failing runs of the unit test PR check

Comment (plkokanov) [reply pkg/gardenlet/controller/backupentry/reconciler.go]: I should have been more specific. The initial finding was from https://pkg.go.dev/golang.org/x/tools/go/analysis/passes/nilness#cond

However, if you just remove the nil check and leave it like:
```
if err := r.SeedClient.Get(seedCtx, client.ObjectKeyFromObject(extensionBackupEntry), extensionBackupEntry); err != nil {
...
} else {
...
}
```
`golangci-lint` throws an error: 
```
Executing golangci-lint
pkg/gardenlet/controller/backupentry/reconciler.go:344:3: early-return: if c { ... } else { ... return } can be simplified to if !c { ... return } ... (move short variable declaration to its own line if necessary) (revive)
                if err := r.SeedClient.Get(seedCtx, client.ObjectKeyFromObject(extensionBackupEntry), extensionBackupEntry); err != nil {
                        if !apierrors.IsNotFound(err) {
                                return reconcile.Result{}, err
                        }
                } else {
                        if lastError := extensionBackupEntry.Status.LastError; lastError != nil {
                                r.Recorder.Event(backupEntry, corev1.EventTypeWarning, gardencorev1beta1.EventDeleteError, lastError.Description)

                                if updateErr := r.updateBackupEntryStatusError(gardenCtx, backupEntry, operationType, lastError.Description, lastError); updateErr != nil {
                                        return reconcile.Result{}, fmt.Errorf("could not update status after deletion error: %w", updateErr)
                                }
                                return reconcile.Result{}, errors.New(lastError.Description)
                        }
                        log.Info("Extension BackupEntry not yet deleted", "extensionBackupEntry", client.ObjectKeyFromObject(extensionBackupEntry))
                        return reconcile.Result{RequeueAfter: RequeueDurationWhenResourceDeletionStillPresent}, nil
                }
```

Comment (rfranzke) [reply pkg/gardenlet/controller/backupentry/reconciler.go]: > I should have been more specific. The initial finding was from https://pkg.go.dev/golang.org/x/tools/go/analysis/passes/nilness#cond

Where do we use this? Or is this something you use locally? I'm confused - either we should add this linter to our configuration (so that everybody can see the results and fix them), or you should disable it locally. Either way, I guess it does not belong to this PR?

Comment (plkokanov) [reply pkg/gardenlet/controller/backupentry/reconciler.go]: This is from the go language server `gopls`. You're right that it wouldn't be used by everyone - and I'm not sure if we can enforce it. 
I added the change to the PR because I noticed it while working on the PR and thought the deletion logic would look a bit cleaner. But I can remove it and add it later on if we think it's worth it.

File: docs/operations/control_plane_migration.md (hunk 4 above)
Comment (plkokanov) [reply docs/operations/control_plane_migration.md]: I think that the deletion of the `BackupEntry` resource as a whole would fail because the client won't be able to connect to the storage provider to do the deletion (e.g. for the aws provider - https://github.com/gardener/gardener-extension-provider-aws/blob/cb2d32524766e416dd11644ceb1b631389bb254e/pkg/controller/backupentry/actuator.go#L40-L45). This should be the case even if the backups were manually removed. If we delete the finalizer of the `extensions.BackupEntry` then.

However, I will add one more clarification to ensure that the `source-etcd-backup` has been removed (automatically) before removing the finalizer.

**Review by rfranzke — COMMENTED** [review]
/lgtm

**Review by timuthy — COMMENTED** [review]
/lgtm
/approve
